cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -174,6 +176,14 @@ private void recordProgramSuspendResume(Id.Program program, String pid, String a
       throw new IllegalArgumentException(msg);
     }
 
+    // If RunRecord has a valid adapterName, then it should match the adapter name passed in.
+    if (record.getAdapterName() != null && (!record.getAdapterName().equals(adapter))) {
```

Comment:
Since this logic is in recordProgramStop, I will use realtime adapter scenario:
Say someone starts a realtime adapter and this will start a Worker and register a RunRecord with null as adapterName (this will also change the status of the Worker driver to RUNNING on the UI aka through Program status REST APIs). If someone hits stop on the UI or issues REST call to the Worker Driver program, then this scenario might occur. Though in this case, passed in adapterName will be `null` instead of another adapter name. I can't think of a way where both adapter names are not-null. Also, I can't think of a scenario where RunRecord has a null for adapterName but passed in adapter name is not-null.